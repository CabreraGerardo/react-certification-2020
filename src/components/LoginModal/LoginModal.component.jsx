/* eslint-disable jsx-a11y/label-has-associated-control */
import React, { useState, useContext } from 'react';
import ReactDOM from 'react-dom';
import { AppContext } from '../../providers/appProvider';
import {
  Modal,
  LoginForm,
  CloseButton,
  Input,
  LoginButton,
  ErrorMessage,
} from './LoginModal.styles';

export default function LoginModal({ isOpen, onClose }) {
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const [error, setError] = useState(null);

  const { dispatch } = useContext(AppContext);

  const handleEmailChange = (event) => {
    setEmail(event.target.value);
  };

  const handlePasswordChange = (event) => {
    setPassword(event.target.value);
  };

  const handleLogin = (event) => {
    event.preventDefault();
    if (email.trim() === 'Wizeline' && password.trim() === 'Rocks!') {
      dispatch({
        type: 'LOG_IN',
      });
      onClose();
    } else {
      setError('The email or the password are incorrect');
    }
  };

  return ReactDOM.createPortal(
    <Modal isOpen={isOpen}>
      <h1>Login</h1>
      <LoginForm onSubmit={handleLogin}>
        <label htmlFor="email">
          Email:
          <br />
          <Input
            id="email"
            name="email"
            type="text"
            placeholder="youremail@email.com"
            onChange={handleEmailChange}
            value={email}
            required
          />
        </label>
        <br />
        <label htmlFor="password">
          Password:
          <br />
          <Input
            id="password"
            name="password"
            type="password"
            placeholder="********"
            onChange={handlePasswordChange}
            value={password}
            required
          />
        </label>
        <br />
        <LoginButton type="submit">LogIn</LoginButton>
        <br />
      </LoginForm>
      <ErrorMessage>{error}</ErrorMessage>
      <CloseButton type="button" onClick={onClose}>
        &times;
      </CloseButton>
    </Modal>,
    document.body
  );
}
