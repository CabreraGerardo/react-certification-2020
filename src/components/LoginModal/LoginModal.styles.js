import styled from 'styled-components';

export const Modal = styled.div`
  position: fixed;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  top: 80px;
  right: ${(props) => (props.isOpen ? '50px' : '-600px')};
  padding: 15px 25px;
  padding-top: 0px;
  border-radius: 14px;
  z-index: 5000;
  width: 300px;
  height: max-content;
  background-color: ${(props) => props.theme.itemBackground};
  color: ${(props) => props.theme.fontColor};
  opacity: ${(props) => (props.isOpen ? '100' : '0')};
  transition: 200ms;
`;

export const LoginForm = styled.form`
  width: max-content;
  margin-bottom: 30px;
  justify-content: center;
  align-items: center;
`;

export const CloseButton = styled.button`
  position: absolute;
  top: 15px;
  right: 15px;
  width: 25px;
  height: 25px;
  padding: 5px;
  cursor: pointer;
  border-radius: 50px;
  background-color: ${(props) => props.theme.shadowColor};
  color: ${(props) => props.theme.fontColor};
  border: 0px;
`;

export const Input = styled.input`
  border: 0px;
  border-radius: 100px;
  padding: 6px 12px;
  margin-bottom: 12px;
  background-color: ${(props) => props.theme.inputColor};
`;

export const LoginButton = styled.button`
  width: max-content;
  align-self: center;
  height: 25px;
  padding: 0px 25px;
  cursor: pointer;
  border-radius: 50px;
  background-color: ${(props) => props.theme.hoverColor};
  color: ${(props) => props.theme.fontColor};
  border: 0px;
`;

export const ErrorMessage = styled.small`
  font-size: 10pt;
  max-width: 200px;
  text-align: center;
`;
