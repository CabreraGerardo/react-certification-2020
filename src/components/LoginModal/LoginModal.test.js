import React from 'react';
import ReactDOM from 'react-dom';
import { render } from '@testing-library/react';
import LoginModal from './LoginModal.component';

describe('Modal Component Tests', () => {
  let mockHideModal;
  beforeEach(() => {
    mockHideModal = jest.fn();
  });

  beforeAll(() => {
    ReactDOM.createPortal = jest.fn((element) => {
      return element;
    });
  });

  afterEach(() => {
    ReactDOM.createPortal.mockClear();
  });

  it('Modal renders', () => {
    const { container } = render(
      <LoginModal isOpen onClose={mockHideModal}>
        <p>My Modal!</p>
      </LoginModal>
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
