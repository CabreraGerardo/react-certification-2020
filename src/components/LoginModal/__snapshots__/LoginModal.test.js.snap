// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal Component Tests Modal renders 1`] = `
<div
  class="sc-bdfBwQ ctjBof"
>
  <h1>
    Login
  </h1>
  <form
    class="sc-gsTCUz hxOkJz"
  >
    <label
      for="email"
    >
      Email:
      <br />
      <input
        class="sc-hKgILt ihGdRK"
        id="email"
        name="email"
        placeholder="youremail@email.com"
        required=""
        type="text"
        value=""
      />
    </label>
    <br />
    <label
      for="password"
    >
      Password:
      <br />
      <input
        class="sc-hKgILt ihGdRK"
        id="password"
        name="password"
        placeholder="********"
        required=""
        type="password"
        value=""
      />
    </label>
    <br />
    <button
      class="sc-eCssSg bcXVum"
      type="submit"
    >
      LogIn
    </button>
    <br />
  </form>
  <small
    class="sc-jSgupP hWTVzv"
  />
  <button
    class="sc-dlfnbm fKCcHc"
    type="button"
  >
    Ã—
  </button>
</div>
`;
